### PR TITLE
feat: add nw.Object dtype

### LINKED ISSUE #343: add nw.Object dtype
This shouldn't do anything clever

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #343 

## What type of PR is this? (check all applicable)

- [ ] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Related issues 

- Related issue # 
- Closes #

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

### CODE REVIEW

File: tests/frame/schema_test.py [hunk 1]
```diff
@@ -12,3 +17,25 @@
...
     assert {"a": nw.Int32()} == {"a": nw.Int32()}
+
+
+def test_object() -> None:
+    df = pd.DataFrame({"a": [1, 2, 3]}).astype(object)
+    result = nw.from_native(df).schema
+    assert result["a"] == nw.Object
+
+
+def test_string_disguised_as_object() -> None:
```
Comment (FBruzzesi) [inline tests/frame/schema_test.py:28]: Love the naming!

File: narwhals/_pandas_like/utils.py [hunk 2]
```diff
@@ -412,7 +413,15 @@ def translate_dtype(dtype: Any) -> DType:
...
-        return dtypes.String()
+        if (idx := column.first_valid_index()) is not None and isinstance(
+            column.loc[idx], str
+        ):
+            # Infer based on first non-missing value.
+            # For pandas pre 3.0, this isn't perfect.
+            # After pandas 3.0, pandas has a dedicated string dtype
+            # which is inferred by default.
+            return dtypes.String()
+        return dtypes.Object()
```
Comment (FBruzzesi) [inline narwhals/_pandas_like/utils.py:424]: Not sure if the comment is related, but shouldn't all non-null/nan elements be str?

**Review by FBruzzesi — COMMENTED** [review]
Just noticed it is already merged! I left one comment 🤔

File: narwhals/_pandas_like/utils.py (hunk 2 above)
Comment (MarcoGorelli) [reply narwhals/_pandas_like/utils.py:424]: you can store anything in object dtype, it doesn't have to be strings

Comment (FBruzzesi) [reply narwhals/_pandas_like/utils.py:424]: That's exactly my point, _all_ elements should be str, not just the first one.

Reproducible example with the above implementation:

```py
class Bar:
    ...

df = pd.DataFrame({"a": ["foo", Bar()]}).astype(object)
result = nw.from_native(df).schema
result["a"] == nw.String
```

```terminal
True
```

Comment (MarcoGorelli) [reply narwhals/_pandas_like/utils.py:424]: ah sure, I see what you mean

true. this is just an approximation, we're just matching what Polars does here https://docs.pola.rs/api/python/stable/reference/series/index.html

**Pull request merged**